### PR TITLE
Fix bug while merging `RecordBatch`, add `SortPreservingMerge` fuzz tester

### LINKED ISSUE #1676: Regression: Merge assertion:  `'assertion failed: i < self.len()'` in array_primitive.rs while merging non overlapping streams
**Describe the bug**
If we try to merge two streams that are non overlapping (e.g. `0..20` and `20..40`) then the following panic is produced:
```
thread 'test_merge_2_no_overlap' panicked at 'assertion failed: i < self.len()', /Users/alamb/.cargo/registry/src/github.com-1ecc6299db9ec823/arrow-8.0.0/src/array/array_primitive.rs:120:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Note this won't affect any users of DataFusion unless they are using externalized sort that @yjshen  has recently been working on or directly use `SortPreservingMergeStream` as IOx does

**To Reproduce**
See fuzztester PR: https://github.com/apache/arrow-datafusion/pull/1678

**Expected behavior**
Correct answers come out, with no panic

**Additional context**
Found in some of our tests in IOx: https://github.com/influxdata/influxdb_iox/issues/3530

I am fairly confident it is a corner case in the code from https://github.com/apache/arrow-datafusion/pull/1596

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Fixes https://github.com/apache/arrow-datafusion/issues/1676 

Includes the code from @yjshen https://github.com/alamb/arrow-datafusion/pull/4 in a separate commit ❤️ 

 # Rationale for this change
Despite pretty good testing in `SortPreservingMerge` there was at least one case that is not yet covered.  This PR adds 

# What changes are included in this PR?
1. A "fuzz" tester that tests merging various combinations of streams (broken into multiple batches, as well as various ranges)
2. Includes the code from @yjshen https://github.com/alamb/arrow-datafusion/pull/4 in a separate commit that fixes #1676 

# Are there any user-facing changes?
Bug is fixed

### CODE REVIEW

**Comment by alamb** [comment]
Marking as a draft as I still plan to add a few more cases to this tester (as I work on #1676 )

File: datafusion/src/physical_plan/sorts/sort_preserving_merge.rs [hunk 1]
```diff
@@ -410,40 +410,53 @@ impl SortPreservingMergeStream {
             // Cursor is not finished - don't need a new RecordBatch yet
             return Poll::Ready(Ok(()));
         }
-        let mut streams = self.streams.streams.lock().unwrap();
+        let mut empty_batch = false;
```
Comment (alamb) [inline datafusion/src/physical_plan/sorts/sort_preserving_merge.rs:413]: The changes to this file are easier to see if you use a whitespace blind diff:
https://github.com/apache/arrow-datafusion/pull/1678/files?w=1

File: datafusion/tests/merge_fuzz.rs [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        // (0..100)
+        // (0..100)
+        make_staggered_batches(0, 100, 2),
+        make_staggered_batches(0, 100, 3),
+    ])
+    .await
+}
+
+#[tokio::test]
+async fn test_merge_2_no_overlap() {
```
Comment (alamb) [inline datafusion/tests/merge_fuzz.rs:49]: this test fails without the changes to `datafusion/src/physical_plan/sorts/sort_preserving_merge.rs`

**Review by houqp — APPROVED** [review]
LGTM
